File: openwrt-yun-bridge/yunbridge/rpc/crc.py
"""CRC-32 IEEE 802.3 implementation for frame integrity verification.

[SIL-2 COMPLIANCE]
This module provides CRC32 calculation that is bit-identical to the
MCU-side implementation in protocol/crc.cpp. Both use:
- Polynomial: 0xEDB88320 (bit-reversed IEEE 802.3)
- Initial value: 0xFFFFFFFF
- Final XOR: 0xFFFFFFFF

The Python stdlib binascii.crc32() uses the same algorithm.

Example:
    >>> from yunbridge.rpc.crc import crc32_ieee
    >>> crc32_ieee(b"123456789")  # Should return 0xCBF43926
    3421780262
"""

from __future__ import annotations

from binascii import crc32

from yunbridge.rpc.protocol import CRC32_MASK


def crc32_ieee(data: bytes, initial: int = 0) -> int:
    """Compute IEEE 802.3 CRC-32 checksum.

    This function wraps Python's binascii.crc32() and masks the result
    to ensure consistent 32-bit unsigned output.

    Args:
        data: Input bytes to checksum.
        initial: Initial CRC value (default 0, internally converted).

    Returns:
        32-bit unsigned CRC value masked by CRC32_MASK.

    Note:
        The result is bit-identical to crc32_ieee() in protocol/crc.cpp
        on the MCU side.
    """
    return crc32(data, initial) & CRC32_MASK

File: openwrt-library-arduino/src/protocol/crc.cpp
/**
 * @file crc.cpp
 * @brief CRC32 implementation for frame integrity verification.
 * 
 * [SIL-2 COMPLIANCE]
 * This module provides a self-contained CRC32 implementation to ensure:
 * - Bit-identical results between MCU (C++) and daemon (Python)
 * - No external library dependencies that could drift
 * - Deterministic execution (no heap, no recursion)
 * 
 * Algorithm: IEEE 802.3 CRC32 (same as Ethernet, PNG, ZIP)
 * Polynomial: 0xEDB88320 (bit-reversed representation)
 * Initial value: 0xFFFFFFFF
 * Final XOR: 0xFFFFFFFF (via bitwise NOT)
 * 
 * @param data Pointer to data buffer
 * @param len  Length of data in bytes
 * @return CRC32 checksum (32-bit unsigned)
 */
#include "crc.h"
#include "rpc_protocol.h"

// Removed dependency on external CRC32 library to ensure consistency
// and avoid potential library version mismatches.
// Implements standard IEEE 802.3 CRC32 (polynomial rpc::RPC_CRC_POLYNOMIAL reversed).

uint32_t crc32_ieee(const uint8_t* data, size_t len) {
  uint32_t crc = rpc::RPC_CRC_INITIAL;
  for (size_t i = 0; i < len; i++) {
    crc ^= data[i];
    for (uint8_t j = 0; j < 8; j++) {
      if (crc & 1) {
        crc = (crc >> 1) ^ rpc::RPC_CRC_POLYNOMIAL;
      } else {
        crc = (crc >> 1);
      }
    }
  }
  return ~crc;
}